core/client/src/main/java/alluxio/client/block/RetryHandlingBlockWorkerClient.java

```diff
@@ -55,47 +58,58 @@
 public final class RetryHandlingBlockWorkerClient
     extends AbstractThriftClient<BlockWorkerClientService.Client> implements BlockWorkerClient {
   private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+  private static final ScheduledExecutorService HEARTBEAT_POOL = Executors.newScheduledThreadPool(
+      Configuration.getInt(PropertyKey.USER_BLOCK_WORKER_CLIENT_THREADS),
+      ThreadFactoryUtils.build("block-worker-heartbeat-%d", true));
+  private static final ExecutorService HEARTBEAT_CANCEL_POOL = Executors.newFixedThreadPool(5,
+      ThreadFactoryUtils.build("block-worker-heartbeat-cancel-%d", true));
 
   private final Long mSessionId;
   // This is the address of the data server on the worker.
   private final InetSocketAddress mWorkerDataServerAddress;
   private final WorkerNetAddress mWorkerNetAddress;
   private final InetSocketAddress mRpcAddress;
-  private final ExecutorService mExecutorService;
-  private final HeartbeatExecutor mHeartbeatExecutor;
-  private final Future<?> mHeartbeat;
+
+  private ScheduledFuture<?> mHeartbeat = null;
 
   /**
    * Creates a {@link RetryHandlingBlockWorkerClient}. Set sessionId to null if no session Id is
    * required when using this client. For example, if you only call RPCs like promote, a session
    * Id is not required.
    *
    * @param workerNetAddress to worker's location
-   * @param executorService the executor service
    * @param sessionId the id of the session
    * @throws IOException if it fails to register the session with the worker specified
    */
-  public RetryHandlingBlockWorkerClient(WorkerNetAddress workerNetAddress,
-      ExecutorService executorService, Long sessionId) throws IOException {
+  public RetryHandlingBlockWorkerClient(WorkerNetAddress workerNetAddress, final Long sessionId)
+      throws IOException {
     mRpcAddress = NetworkAddressUtils.getRpcPortSocketAddress(workerNetAddress);
 
     mWorkerNetAddress = Preconditions.checkNotNull(workerNetAddress);
     mWorkerDataServerAddress = NetworkAddressUtils.getDataPortSocketAddress(workerNetAddress);
-    mExecutorService = executorService;
     mSessionId = sessionId;
-    if (sessionId != null && executorService != null) {
-      mHeartbeatExecutor = new BlockWorkerClientHeartbeatExecutor(this);
+    if (sessionId != null) {
+      // The heartbeat is scheduled to run in a fixed rate. The heartbeat won't consume a thread
+      // from the pool while it is not running.
+      mHeartbeat = HEARTBEAT_POOL.scheduleAtFixedRate(new Runnable() {
+            @Override
+            public void run() {
+              try {
+                sessionHeartbeat();
+              } catch (InterruptedException e) {
+                LOG.info("Heartbeat for session {} is " + "interrupted" + ".", sessionId);
```

Comment:
Not sure if we need to log, there isn't any significance of this situation right?